### PR TITLE
First-pass for player collision improvements

### PR DESCRIPTION
**Pull request by rgarmsen2295**
Hey all, this implements a first pass of the wall sliding we've needed for a while.  Not perfect as you can still get "sticky" once you're sliding on the wall occasionally, but it's a huge improvement from my testing.

BIG ISSUE though, that, hopefully, you all can help with.

You should see the "hack" comment in BoundingBox.cpp; for some reason the reflection logic was returning x-normal collisions as y-normal collisions.  This is really odd obviously (but does explain why the cookies were going through some hedges but not others, any hedges placed parallel to the x axis were fudged).  This is kind of hard to verbally explain, but if you try throwing cookies through the hedge behind the player at start you'll see what I'm talking about (without this PR that is!).

Anyway, been looking at this for a while and can't see it easily, hopefully, just missing something dumb from the existing reflection or BB logic you all can spot (as I didn't change anything there other than the hack).  Works as is right now but it's the literal definition of a hack :P

### CODE REVIEW

File: src/BoundingBox.cpp [hunk 1]
```diff
@@ -47,7 +47,9 @@ glm::vec3 BoundingBox::calcReflNormal(BoundingBox& other) {
    }
    if((this->min_.y - epsilon <= other.max_.y && this->min_.y + epsilon >= other.max_.y) ||
            (this->max_.y - epsilon <= other.min_.y && this->max_.y + epsilon >= other.min_.y)){
-       normal = glm::vec3(0.0, 1.0, 0.0);
+
+       // TODO(rgarmsen2295): Hack, should be y as normal, figure out why collisions are matching x instead of y only
+       normal = glm::vec3(1.0, 0.0, 0.0);
```
Comment (joncatanio) [inline src/BoundingBox.cpp]: Could the hedges be rotated (or lack rotation)? Do we know the initial orientation of `cube.obj` it's possible it is rotated 90 degrees along the z-axis or something? I dunno if this would actually effect anything if we are computing the `y` value regardless or object orientation. I didn't really play around with the bounding box stuff.

File: src/BunnyPhysicsComponent.cpp [hunk 2]
```diff
@@ -19,11 +19,12 @@ void BunnyPhysicsComponent::initObjectPhysics() {
 	setupInitialRotation();
 	updateBoundingBox();
 
-	std::shared_ptr<GameObject> objHit = world.checkCollision(holder_);
+	std::vector<std::shared_ptr<GameObject>> objsHit = world.checkCollision(holder_);
```
Comment (joncatanio) [inline src/BunnyPhysicsComponent.cpp:22]: And so it begins!

**Review by joncatanio — APPROVED** [review]
Looks good, the feel on the walls is awesome but when I hit a corner it gets stuck and sometimes goes through the wall still. I believe this might be poor bounding box calculations or the multiple object collision test isn't doing exactly what we want. We should try to talk this over tomorrow.

File: src/BoundingBox.cpp (hunk 1 above)
Comment (rgarmsen2295) [reply src/BoundingBox.cpp]: Unless you changed something in the JSON version, the hedges aren't rotated, they're just scaled differently (longer in x vs longer in z).  I thought about that as being part of the problem but didn't see anything immediate in the code pointing to an issue.

As far as the cube.obj file itself having an odd initial rotation, would it matter? The min and max are calculated based off the min and max vertices of the object on load (and then recalculated as the object is transformed), so while worth looking into I can't think of anything off the bat.

File: src/BunnyPhysicsComponent.cpp (hunk 2 above)
Comment (rgarmsen2295) [reply src/BunnyPhysicsComponent.cpp:22]: dun dun dunnnnn lol

File: src/BoundingBox.cpp (hunk 1 above)
Comment (joncatanio) [reply src/BoundingBox.cpp]: nope, I don't think it would matter. Just spit ballin' ideas.

**Comment by rgarmsen2295** [comment]
Happy to go over it tomorrow, I hadn't been able to go through corners so I'm curious to see where/how it happens and thinking about fixing it.

My guess is that it has to do with when you rotate the car - that action changes the transform and therefore the bounding box, but isn't accounted for in the update loop (you're just rotating how the car is drawn after all!), only in the input component.  Therefore, you can be in a position where you rotate into a collision state, move forward and then are placed back into a collision state, which is where things go bad.

There's probably a way out of it, but object rotation has been a sore spot in the code since lab 1.  Since we handle rotations immediately on input, it may be a tad harder to refactor, but, hopefully, not impossible (aka a fat time sink - nothing impossible!).

tl;dr Rotations are the root problem from my guesstimate and we will need to deal with them in the update loop as object movement instead of exclusively in the input controller as they are technically an "update" of the object and beef up the collision resolution accordingly.

File: src/BoundingBox.cpp (hunk 1 above)
Comment (rgarmsen2295) [reply src/BoundingBox.cpp]: Nah, I appreciate it. The problem doesn't make sense, so it's likely to be a non-sensical root issue like that.  Keep spittin' ballin' ideas like you Ballmer on the Ballmer peak, broski xD

**Comment by rgarmsen2295** [comment]
Okay made some changes (see the last commit):

+ Fixed hack
+ Fixed corner slide thru on convex corners
+ Moved/removed two trees to prevent unintentional collision
+ Put mouse movement behind an ifdef (since both players hands should be on the sticks or wasd/ijkl)_

Still trying to figure out why *collision fires early on the ends of hedges* (it's not because they're merged on top of each other like I thought, something else is wonk like you thought @joncatanio).  I also tried changing the car model in case it was just screwy but that didn't help.

Take a look at the new changes and then (assuming they're good) I'll merge this for now and keep thinking about the corner problem (cookies are fine too, new theory might be an issue with the fact that the player's bounding box is bigger than the object it's colliding with but I'm not seeing the problem inherent with that yet, so maybe not).

**Comment by rgarmsen2295** [comment]
I also tried getting fancier with how rotations work in cases where they should cause collision resolution but it got ugly fast w/o a noticeable benefit.  I may look into this again later, but the edge issue is a bigger problem then the minor sticking remaining imo.

**Comment by joncatanio** [comment]
Looks good to me Reed, I'ma go ahead and merge this and we can fix anything else in a new PR. For now this is working and a huge improvement to the collisions on master so lets get this merged.
